packages/ketcher-react/src/script/editor/operations/text.ts

```diff
@@ -0,0 +1,189 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+
+import { BaseOperation } from './base'
+import { OperationType } from './OperationType'
+import { Text, Vec2, scale } from 'ketcher-core'
+import Restruct, { ReText } from '../../render/restruct'
+
+interface TextCreateData {
+  id?: any
+  label: string
+  position: Vec2
+  type: string
+}
+
+export class TextCreate extends BaseOperation {
+  data: TextCreateData
+  performed: boolean
+
+  constructor(id: any, label: string = '', position: Vec2, type: string) {
+    super(OperationType.TEXT_CREATE)
+    this.data = { id, label, position, type }
+    this.performed = false
+  }
+
+  execute(restruct: Restruct): void {
+    const struct = restruct.molecule
+
+    if (!this.performed) {
+      this.data.id = struct.texts.add(new Text(this.data))
+      this.performed = true
+    } else {
+      struct.texts.set(this.data.id, new Text(this.data))
+    }
+
+    restruct.texts.set(this.data.id, new ReText(struct.texts.get(this.data.id)))
+
+    const { id, position } = this.data
+
+    struct.textSetPosition(id, new Vec2(position))
+
+    BaseOperation.invalidateItem(restruct, 'texts', this.data.id, 1)
+  }
+
+  invert(): BaseOperation {
+    return new TextDelete(this.data.id)
+  }
+}
+
+interface TextUpdateData {
+  id?: any
+  label?: string
+  invertedLabel?: string
+  position?: Vec2
+  type?: string
+}
+
+export class TextUpdate extends BaseOperation {
+  data: TextUpdateData
+
+  constructor(
+    id?: any,
+    label?: string,
+    invertedLabel?: string,
+    position?: Vec2,
+    type?: string
+  ) {
+    super(OperationType.TEXT_UPDATE)
+    this.data = { id, label, invertedLabel, position, type }
+  }
+
+  execute(restruct: Restruct) {
+    const { id, label } = this.data
+    const text = restruct.molecule.texts.get(id)
+
+    if (text) {
+      text.label = label
+    }
+
+    BaseOperation.invalidateItem(restruct, 'texts', id, 1)
+  }
+
+  invert() {
+    const inverted = new TextUpdate(this.data.id)
+    inverted.data.label = this.data.invertedLabel
+    inverted.data.invertedLabel = this.data.label
+    return inverted
+  }
+}
+
+interface TextDeleteData {
+  id: any
+  label?: string
+  position?: Vec2
+  type?: string
+}
+
+export class TextDelete extends BaseOperation {
+  data: TextDeleteData
+  performed: boolean
+
+  constructor(id: any) {
+    super(OperationType.TEXT_DELETE)
+    this.data = { id }
+    this.performed = false
+  }
+
+  execute(restruct: Restruct): void {
+    const struct = restruct.molecule
+
+    if (!this.performed) {
+      const item = struct.texts.get(this.data.id)!
```

Comment:
again, you can't garant here that `item` exists, so you should check it for undefined/null

Reply:
fixed